tests/cxx/instance_preds/Shape.cpp

```diff
@@ -0,0 +1,237 @@
+  //@ ensures valid(x, y);
+  {
+    setX(x);
+    setY(y);
+  }
+
+  void moveTo(const Pos &p)
+  //@ requires valid(_, _) &*& p.valid(?x, ?y);
+  //@ ensures valid(x, y) &*& p.valid(x, y);
+  {
+    int px = p.getX();
+    int py = p.getY();
+    moveTo(px, py);
+  }
+};
+
+class Shape
+{
+  Pos m_pos;
+
+public:
+  /*@
+    predicate valid(int x, int y) =
+      struct_Pos_padding(&this->m_pos) &*&
+      (&this->m_pos)->valid(x, y);
+  @*/
+
+  Shape(const Pos &pos) : m_pos(pos)
+  //@ requires pos.valid(?x, ?y);
+  //@ ensures pos.valid(x, y) &*& valid(x, y) &*& Shape_vtype(this, thisType);
+  {
+    //@ close valid(x, y);
+  }
+
+  virtual ~Shape()
+  //@ requires valid(_, _) &*& Shape_vtype(this, thisType);
+  //@ ensures true;
+  {
+    //@ open valid(_, _);
+  }
+
+  virtual int calcArea() const = 0;
+  //@ requires valid(?x, ?y);
+  //@ ensures valid(x, y) &*& result >= 0;
+};
+
+class Square : public Shape
+{
+  int m_width;
+
+public:
+  /*@
+    predicate valid(int x, int y) =
+      this->valid(&typeid(Shape))(x, y) &*&
+      this->m_width |-> ?w &*&
+      w >= 0 &*&
+      Square_bases_constructed(this);
+  @*/
+
+  Square(const Pos &pos, int width) : Shape(pos), m_width(width)
+  //@ requires pos.valid(?x, ?y) &*& width >= 0;
+  //@ ensures valid(x, y) &*& pos.valid(x, y) &*& Square_vtype(this, thisType);
+  {
+    //@ close valid(x, y);
+  }
+
+  ~Square()
+  //@ requires valid(_, _) &*& Square_vtype(this, thisType);
+  //@ ensures true;
+  {
+    //@ open valid(_, _);
+  }
+
+  int calcArea() const override
+  //@ requires valid(?x, ?y);
+  //@ ensures valid(x, y) &*& result >= 0;
+  {
+    //@ open valid(x, y);
+    return m_width * m_width;
+    //@ mul_mono_l(0, this->m_width, this->m_width);
+    //@ close valid(x, y);
+  }
+};
+
+class Rectangle : public Shape
+{
+  int m_width;
+  int m_height;
+
+public:
+  /*@
+    predicate valid(int x, int y) =
+      this->valid(&typeid(Shape))(x, y) &*&
+      this->m_width |-> ?w &*& w >= 0 &*&
+      this->m_height |-> ?h &*& h >= 0 &*&
+      Rectangle_bases_constructed(this);
+  @*/
+
+  Rectangle(const Pos &pos, int width, int height) : Shape(pos), m_width(width), m_height(height)
+  //@ requires pos.valid(?x, ?y) &*& width >= 0 &*& height >= 0;
+  //@ ensures pos.valid(x, y) &*& valid(x, y) &*& Rectangle_vtype(this, thisType);
+  {
+    //@ close valid(x, y);
+  }
+
+  ~Rectangle()
+  //@ requires valid(_, _) &*& Rectangle_vtype(this, thisType);
+  //@ ensures true;
+  {
+    //@ open valid(_, _);
+  }
+
+  int calcArea() const override
+  //@ requires valid(?x, ?y);
+  //@ ensures valid(x, y) &*& result >= 0;
+  {
+    //@ open valid(x, y);
+    return m_width * m_height;
+    //@ mul_mono_l(0, this->m_width, this->m_height);
+    //@ close valid(x, y);
+  }
+};
+
+int calcArea(const Shape &s)
+//@ requires s.valid(?x, ?y);
+//@ ensures s.valid(x, y);
+{
+  return s.calcArea();
+}
+
+void destroy(Shape *s)
+//@ requires s->valid(_, _) &*& new_block_Shape(s);
```

Comment:
I did not adjust `new` and `delete`. New'ing a Square produces a `new_block_Square`. `delete p` calls `p`'s destructor and consumes its `P_new_block` afterwards. Calling a virtual destructor on `p` first consumes and reproduces P_vtype(p, ?p_type) to retrieve `p`'s `vtype`. Then, the destructor is called with `thisType = p_type`. This means that `destroy` in the example can only be used on an object with most derived type `Shape`.

Does it make sense to remove the prefix from `prefix_new_block` and have `new_block(void *p, std::type_info *p_type` instead? New'ing `S` would then call the constructor of `S` and produce `new_block(s, &typeid(S))`. `delete s` then first consumes `new_block(s, _)` and afterwards calls the destructor. I think this would make deleting objects with multiple inheritance work too.

Explicitly calling a destructor only makes sense when placement new (which we don' support yet) was used. We currently do not support calling a destructor explicitly (I should make the error message in that case more clear).

Reply:
> New'ing a Square produces a new_block_Square.

OK, then there is no unsoundness. I suggest we leave polymorphic destruction for next year.

> This means that destroy in the example can only be used on an object with most derived type Shape.

That means this function is pointless since Shape is abstract. Let's just remove it.

 Your two-arg `new_block` suggestion seems like a good idea! Let's do it next year.

Reply:
> OK, then there is no unsoundness.

Actually, this does not resolve the subtlety of the soundness argument.  If `delete` would check that the vtype equals the most derived type, then it would be obviously sound. I suggest we leave this issue for next year. (But if you prefer not to write text about a design that will change, feel free to do it now.)